File: src/main/scala/MyTest/mmlTests/a201401/Scala/gpt35turbo/FibonacciAcceptorImpl.scala
//package MyTest.mmlTests.a201401.Scala.gpt35turbo
//
//import MyTest.mmlTests.a201401.Scala.FibonacciAcceptor
//import scala.collection.mutable.Map
//
//class FibonacciAcceptorImpl extends FibonacciAcceptor {
//
//  private var currentSequence: List[Long] = List.empty
//  private var allSequences: Map[String, List[Long]] = Map.empty
//
//  def reset(sequenceName: String): Unit = {
//    if (allSequences.contains(sequenceName)) {
//      throw new IllegalArgumentException("Cannot create a sequence with existing name!")
//    }
//    currentSequence = List.empty
//    allSequences += (sequenceName -> currentSequence)
//  }
//
//  def consumeNext(l: Long): Boolean = {
//    if (currentSequence.isEmpty) {
//      currentSequence :+= l
//      true
//    } else {
//      val lastIndex = currentSequence.size - 1
//      val secondLast = currentSequence(lastIndex - 1)
//      val last = currentSequence(lastIndex)
//      if (l == secondLast + last) {
//        currentSequence :+= l
//        true
//      } else {
//        false
//      }
//    }
//  }
//
//  def getCurrentSequence: List[Long] = {
//    if (currentSequence.isEmpty) {
//      throw new IllegalStateException("Should reset first!")
//    } else {
//      currentSequence
//    }
//  }
//
//  def getAllSequences: Map[String, List[Long]] = {
//    allSequences.map(entry => entry._1 -> entry._2.toList).toMap
//  }
//}

File: src/main/scala/MyTest/mmlTests/a201401/Java/gpt35turbo/FibonacciAcceptorImpl.java
package MyTest.mmlTests.a201401.Java.gpt35turbo;

import java.util.*;
import MyTest.mmlTests.a201401.Java.FibonacciAcceptor;

public class FibonacciAcceptorImpl implements FibonacciAcceptor {

    private Map<String, List<Long>> sequences;
    private List<Long> currentSequence;

    public FibonacciAcceptorImpl() {
        sequences = new HashMap<>();
        currentSequence = null;
    }

    @Override
    public void reset(String sequenceName) {
        if (sequences.containsKey(sequenceName)) {
            throw new IllegalArgumentException("Sequence name already exists");
        }
        currentSequence = new ArrayList<>();
        sequences.put(sequenceName, currentSequence);
    }

    @Override
    public boolean consumeNext(long l) {
        if (currentSequence == null) {
            throw new IllegalStateException("No reset was initially done");
        }

        if (isFibonacciNumber(l)) {
            currentSequence.add(l);
            return true;
        }
        return false;
    }

    @Override
    public List<Long> getCurrentSequence() {
        if (currentSequence == null) {
            throw new IllegalStateException("No reset was initially done");
        }
        return currentSequence;
    }

    @Override
    public Map<String, List<Long>> getAllSequences() {
        return new HashMap<>(sequences);
    }

    private boolean isFibonacciNumber(long num) {
        if (currentSequence.isEmpty() || currentSequence.size() == 1) {
            return true;
        }

        long secondLast = currentSequence.get(currentSequence.size() - 2);
        long last = currentSequence.get(currentSequence.size() - 1);

        return num == (last + secondLast);
    }
}